engineioxide/src/service.rs

```diff
@@ -222,8 +247,32 @@ impl RequestInfo {
     /// Parse the request URI to extract the [`TransportType`](crate::service::TransportType) and the socket id.
     fn parse<B>(req: &Request<B>) -> Result<Self, Error> {
         let query = req.uri().query().ok_or(UnknownTransport)?;
-        if !query.contains("EIO=4") {
-            return Err(UnsupportedProtocolVersion);
+
+        let protocol: ProtocolVersion = query
+            .split('&')
+            .find(|s| s.starts_with("EIO="))
+            .and_then(|s| s.split('=').nth(1))
+            .ok_or(UnknownTransport)
+            .and_then(|t| t.parse())?;
+
+        cfg_if! {
+            if #[cfg(all(feature = "v3", feature = "v4"))] {
+                if protocol != ProtocolVersion::V3 && protocol != ProtocolVersion::V4 {
+                    return Err(Error::UnsupportedProtocolVersion);
+                }
+            }       
+            else if #[cfg(feature = "v4")] {
+                if protocol != ProtocolVersion::V4 {
+                    return Err(Error::UnsupportedProtocolVersion);
+                }
+            }
+            else if #[cfg(feature = "v3")] {
+                if protocol != ProtocolVersion::V3 {
+                    return Err(Error::UnsupportedProtocolVersion);
+                }
+            } else {
```

Comment:
This could go to the `ProtocolVersion` from_str impl

Reply:
Something like this:
```rust
#[cfg(all(feature = "v3", feature = "v4"))]
fn from_str(s: &str) -> Result<Self, Self::Err> {
    match s {
        "3" => Ok(ProtocolVersion::V3),
        "4" => Ok(ProtocolVersion::V4),
        _ => Err(Error::UnsupportedProtocolVersion),
    }
}

#[cfg(feature = "v4")]
fn from_str(s: &str) -> Result<Self, Self::Err> {
    match s {
        "4" => Ok(ProtocolVersion::V4),
        _ => Err(Error::UnsupportedProtocolVersion),
    }
}

#[cfg(feature = "v3")]
fn from_str(s: &str) -> Result<Self, Self::Err> {
    match s {
        "3" => Ok(ProtocolVersion::V3),
        _ => Err(Error::UnsupportedProtocolVersion),
    }
}
```

Reply:
Right! I love this, it's much cleaner.